Update Butler matching notebook for Run 1.2i

jiwoncpark opened the pull request:
As discussed with @jchiang87, I modified the `matching_stack.ipynb` tutorial to reflect the changes that came with Run 1.2i. I also added some annotations at points where I got stuck changing the code to work for Run 1.2i.

Update summary:
- Explains how to extract information about the repo from Butler, e.g. through `getKeys()` and `queryMetadata()`
- Explains how to get available filter, tract, patch combinations
- Updates the repo locations for src and deepcoadd in Jupyter-dev
- Updates the src dataset key schema for Run 1.2i, e.g. `raft` --> `raftName`
- Uses new column names for kronflux

I'm sorry if I missed something obvious, but I can't figure out why the number of matches between the drp catalog (from src) and protoDC2 is 0. I inferred from the original dataId in Jim's notebook that `{'visit': 219976, 'raft': '2,2', 'sensor': '1,1'}` for Run 1.1p were the same as `{'visit': 219976, 'raftName': 'R22', 'detectorName': 'S11'}` for Run 1.2i. Was I wrong?

fjaviersanchez commented:
@jiwoncpark, I think that you can try to match directly to `dc2_truth_run1.2_static` so you don't have to worry about the rotation of the catalog. It is not finding any matches with the `src_catalog` because the regions don't overlap. You probably want to query a different sensor-raft (blue points are from the `obj_catalog` and orange are from the `src_catalog`):

![image](https://user-images.githubusercontent.com/13971732/53590626-13a7ed80-3b47-11e9-947b-c7aa65ddf48a.png)

fjaviersanchez commented:
Another option is to query a different tract, patch (you can check which ones overlap with the visit in the notebook using: `overlaps[(overlaps['visit']==219976) & (overlaps['detector']==94)]`. E.g, using:
```
our_filter = 'r'
our_tract = 4639
our_patch = '6,6'
```

You get ~350 matches

jiwoncpark commented:
Hi @fjaviersanchez, thank you for looking into this! Are you suggesting that I work with the truth catalog instead, which would entail turning it into a `afw_table` without the rotation and passing it into the `matchRaDec` function? If I were to stick with protoDC2, how would I go about figuring out which sensor, raft combo overlaps with a given region in protoDC2?

fjaviersanchez commented:
You can use either `protoDC2` or `dc2_truth` indistictively. The only difference would be that you don't have to rotate (you would still generate the afw.table/SourceCatalog object with `dc2_truth` as you said and in the function where you do the selection you avoid the rotation step). 

The non-matching is a separate problem. In order to figure out which object catalog (coadd) overlaps with a given exposure (source catalog). In the `overlaps` table in your notebook you have all the information that you need (`R22, S11` is `detector=94`).

This is because in this line:
`galaxy_catalog = region_selector(gc, band=our_filter, max_mag=mag_max)`

You get rid of anything out of any galaxies lying outside of the patch that you queried (that's why you weren't finding any matches in the src catalog, because the piece of `protoDC2` that you had didn't overlap with `src`).

rmandelb commented:
@jiwoncpark - I just wanted to check in about this PR, since there hasn't been activity for a while.  My impression from your exchange with @fjaviersanchez was that you may be planning additional changes before it's ready for review, so I was waiting on that - but please let me know if that's incorrect and I can seek out a reviewer now.  Thanks!

reneehlozek commented:
hi @jiwoncpark, just to check again, are there additional changes you are making on this PR? thanks!

jiwoncpark commented:
Oh no, I missed Rachel's post from earlier. I'm so sorry! Yes, I recall I made some changes per suggestions from @fjaviersanchez but it seems I can't access the NERSC Jupyter server now (maintenance?) -- I'll request reviewers by this Thursday.

jiwoncpark commented:
Thank you for your patience! In this version, ready for review, I've 
- implemented @fjaviersanchez 's suggestion of using the truth catalog instead and
- added a method of finding the tract and patch combo for `coadd` that overlaps with a given `src` region, instead of an arbitrarily sky coordinate as I had in the previous version.

jiwoncpark commented:
@wmwv has kindly agreed to review this PR!

jiwoncpark commented:
I've also asked @jchiang87 (who wrote the original tutorial for Run 1.1) to review this!

reneehlozek commented:
Great, good to hear, thanks!

wmwv commented on tutorials/matching_stack_run_1.2i.ipynb:
```diff
@@ -0,0 +1,514 @@
+    "    \"\"\"Initializes an afw table with magnitude column(s)\"\"\"\n",
+    "    def define_mag_columns(bands):\n",
+    "        \"\"\"Defines columns for given list of bands\"\"\"\n",
+    "        Coldef = namedtuple('Coldef', 'name type doc'.split())\n",
```
`Coldef = namedtuple('Coldef', ['name', 'type', 'doc'])

wmwv commented on tutorials/matching_stack_run_1.2i.ipynb:
```diff
@@ -0,0 +1,514 @@
+    "    # Filter is r for our purposes, but we can get it this way too\n",
+    "    observed_filter = calexp.getInfo().getFilter().getName()\n",
+    "    # Get primary extended objects with no flux flag\n",
+    "    flux_model = 'base_PsfFlux' # schema v3\n",
```
Use `slot_PsfFlux` instead of `base_PsfFlux`.

wmwv commented on tutorials/matching_stack_run_1.2i.ipynb:
```diff
@@ -0,0 +1,514 @@
+    "\n",
+    "We can consider visit-level src catalog data, in which case we would define `datasetType='src'` and provide a query, or `dataId`, to the butler with (`visit`, `raftName`, `sensorName`) ids; or we can consider coadd object data, for which we would define `datasetType='coadd'` and provide a `dataId` with (`filter`, `tract`, `patch`) ids.\n",
+    "\n",
+    "Somewhat different flux models are available in the Run1.2i data for src catalogs versus coadd catalogs. We use `'base_PsfFlux'` for the flux model, as they are available in both.\n",
```
Use `slot_PsfFlux` instead of `base_PsfFlux`.
`slot_` is the data column required by the processing
`base_` is how it was computed.
For most cases `slot_` points to `base_` and this doesn't matter.  But conceptually `slot_` is what we want.

wmwv commented on tutorials/matching_stack_run_1.2i.ipynb:
```diff
@@ -0,0 +1,514 @@
+    "\n",
+    "    # Difference in magnitudes vs separation.\n",
+    "    fig.add_subplot(2, 2, 4)\n",
+    "    plt.errorbar(sep, delta_mag, fmt='.')\n",
```
If I match the object to the source catalogs, those matches and magnitude deltas both look reasonable and fine.

wmwv replied:
And, in fact, given the difficulties with the truth catalogs for Run 1.2, it may be worth adding an example of matching between the `coadd` and `src` catalogs to show that all of this really does work.

wmwv commented on tutorials/matching_stack_run_1.2i.ipynb:
```diff
@@ -0,0 +1,514 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def compare_differences(drp_catalog, matches, observed_filter):\n",
```
The `drp_catalog` input isn't necessary.  `matches` contains everything you need to know for this function.

reneehlozek commented:
hi @jiwoncpark - is there an update on the changes requested by @wmwv for this PR?

jiwoncpark commented:
I'll work on updating this for Run 2.xi.

reneehlozek commented:
Perhaps we can hack these final updates based on 2.1/2.2i at the Spring Day @jiwoncpark?

reneehlozek commented:
hi @jiwoncpark, just wanted to check on the status of this notebook?

jiwoncpark commented:
Hi @reneehlozek, I'm not sure if this tutorial is useful anymore, as I see that GCRCatalogs now publishes `dc2_object_run2.1i_dr1*_matched` and people can now just load that. I wrote the tutorial to teach myself how to match to truth, back when these weren't available. Do you think it would still be useful to update Jim's 1.1p matching notebook to 2.1i?

yymao commented:
superseded by #135
